src/config-util/getConfig.js

```diff
@@ -0,0 +1,95 @@
+'use strict'
+
+const _ = require('lodash')
+const defaultConfig = require('./defaultConfig')
+
+/**
+ * Check if the config is "simple" i.e. doesn't contains any of full config keys.
+ *
+ * This check is not concerned with the validity of the config. For example,
+ * en empty object or array will satisfy the check.
+ *
+ * @param config
+ * @returns {boolean}
+ */
+function isSimple(config) {
+  /*
+    Scenarios which satisfy `isSimple` check:
+    - []
+    - {}
+    - [
+        {
+          "includes": ["src/*.js"],
+          "excludes": ["!src/*.ignore.js"],
+          "commands": ["eslint --fix", "git add"]
+        }
+      ]
+    - { "*.js": ["eslint --fix", "git add"] }
+
+    Scenario which do not satisfy the check:
+    - {
+        "linters": { "*.js": ["eslint --fix", "git add"] }
+      }
+    - {
+        "linters": []
+      }
+  */
+  return (
+    Array.isArray(config) ||
+    (_.isObject(config) &&
+      !_.has(config, 'linters') &&
+      _.intersection(Object.keys(defaultConfig), Object.keys(config)).length === 0)
+  )
+}
+
+/**
+ * Expand shorthand object format to the array format. If `linters` is an array, it is
+ * returned as is. If it is a plain object, it is converted from
+ * `{ "*.js": "eslint --fix" }` to
+ * `[ { "includes": ["*.js"], "commands": "eslint --fix" } ]`
+ *
+ * @param {Array|Object|null|undefined} linters
+ */
+function expandShorthands(linters) {
+  if (_.isNil(linters) || _.isEmpty(linters)) return []
+
+  if (Array.isArray(linters)) return linters
+
+  return Object.keys(linters).reduce((expanded, pattern) => {
+    expanded.push({ includes: [pattern], commands: linters[pattern] })
+    return expanded
+  }, [])
+}
+
+/**
+ * For a given configuration object that we retrive from .lintstagedrc or package.json
+ * construct a full configuration with all options set.
+ *
+ * This is a bit tricky since we support 2 different syntxes: simple and full
+ * For simple config, only the `linters` configuration is provided.
+ *
+ * @param {Object} sourceConfig
+ * @returns {{
+ *  concurrent: boolean, chunkSize: number, gitDir: string, globOptions: {matchBase: boolean, dot: boolean}, linters: {}, subTaskConcurrency: number, renderer: string, verbose: boolean
+ * }}
+ */
+module.exports = function getConfig(sourceConfig) {
+  const isSimpleConfig = isSimple(sourceConfig)
+  // At this point, the config hasn't been validated.
+  const linters = expandShorthands(
+    // `_.get` is used because `linters` property might not be present.
+    isSimpleConfig ? sourceConfig : _.get(sourceConfig, 'linters')
+  )
+  const config = _.defaultsDeep(
+    { linters }, // Do not mutate sourceConfig!!!
+    isSimpleConfig ? null : sourceConfig,
+    defaultConfig
+  )
+
+  // Check if renderer is set in sourceConfig and if not, set accordingly to verbose
+  if (_.isObject(sourceConfig) && !_.has(sourceConfig, 'renderer')) {
```

Comment:
Why do you need the first check? JSDoc says it is an `Object`.

Let's rewrite as `sourceConfig.hasOwnProperty('renderer')`?

Reply:
This results in an eslint error - `[eslint] Do not access Object.prototype method 'hasOwnProperty' from target object. (no-prototype-builtins)`.

Reply:
Since we don't use `Object.create(null)` anywhere, I'd disable the rule https://eslint.org/docs/rules/no-prototype-builtins in `.eslintrc`